### PR TITLE
added proj-config-utils and its proof of concept functionality to new-repo

### LINKED ISSUE #57: proj-config-utils
As a developer, I want configuring my project to be easy and saveable.

### PR DESCRIPTION
**Pull request by eetupur**
resolves #57

### CODE REVIEW

**Review by veikkahonkanen — APPROVED** [review]

**Review by veikkahonkanen — APPROVED** [review]

File: src/classes/proj-config-utils.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+export class ProjConfigUtils {
+
+  /**
+   * Reads project-config.json from given path
+   * 
+   * @param outerFolder outer folder of the project (check path-utils for more info)
+   * 
+   * @returns project config as a JSON object that has an interface.
+   */
+    static async readProjConfig(outerFolder : string) : Promise<ProjConfigJson> {
```
Comment (jarinykanen) [inline src/classes/proj-config-utils.ts]:
```suggestion
    static async readProjConfig(outerFolder: string): Promise<ProjConfigJson> {
```

File: src/classes/proj-config-utils.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import fs from "fs";
+import { ProjConfigJson } from "../interfaces/types";
+import path from "path";
+
+const configFileName = "project-config.json";
+const projConfigTemplate = `.${path.sep}resources${path.sep}project-config-template.json`;
+
+export class ProjConfigUtils {
```
Comment (jarinykanen) [inline src/classes/proj-config-utils.ts:11]: Add tsdocs

File: src/classes/proj-config-utils.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  /**
+   * Writes to project-config.json
+   * 
+   * @param path outer folder of the project (check path-utils for more info)
+   * 
+   * @param data should receive the edited version of project-config.json 
+   */
+  static async writeProjConfig(outerFolder : string, data : string) : Promise<void> {
```
Comment (jarinykanen) [inline src/classes/proj-config-utils.ts]:
```suggestion
  static async writeProjConfig(outerFolder: string, data: string): Promise<void> {
```

File: src/classes/proj-config-utils.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    } catch(err) {
+      throw new Error ("Error while reading project-config: " + err);
+    }
+  }
+
+  /**
+   * Writes to project-config.json
+   * 
+   * @param path outer folder of the project (check path-utils for more info)
+   *
```
Comment (jarinykanen) [inline src/classes/proj-config-utils.ts]:
```suggestion
```

File: src/classes/proj-config-utils.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      await this.checkProjConfig(outerFolder);
+
+      fs.writeFileSync(this.outerToFile(outerFolder), data, "utf8");
+    } catch (err) {
+      throw new Error("Error when attempting to write to user-config:" + err);
+    }
+  }
+
+  /**
+   * creates project-config.json from template to given folder
```
Comment (jarinykanen) [inline src/classes/proj-config-utils.ts]:
```suggestion
   * Creates project-config.json from template to given folder
```

File: src/classes/proj-config-utils.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+   * 
+   * @param outerFolder foldern in which project-config.json should be created to
+   */
+  static async createProjConfig(outerFolder : string) : Promise<void> {
```
Comment (jarinykanen) [inline src/classes/proj-config-utils.ts]:
```suggestion
  static async createProjConfig(outerFolder: string): Promise<void> {
```

File: src/classes/proj-config-utils.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      throw new Error("Error when attempting to create user-config:" + err);
+    }
+  }
+
+  /**
+   * Checks if user-config.json exists in the root and creates it from template if it doesn't
+   * 
+   * @param outerFolder the path to project-config.json of the project
+   */
+  private static async checkProjConfig(outerFolder : string) : Promise<void> {
```
Comment (jarinykanen) [inline src/classes/proj-config-utils.ts]:
```suggestion
  private static async checkProjConfig(outerFolder: string): Promise<void> {
```

File: src/classes/proj-config-utils.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+      if (!fs.existsSync(this.outerToFile(outerFolder))) {
+        this.createProjConfig(outerFolder); 
+      }
+    } catch (err) {
+      throw new Error("Error when attempting to create user-config:" + err);
+    }
+  }
+
+  /**
+   * gives the path to project-config.json inside a project folder
```
Comment (jarinykanen) [inline src/classes/proj-config-utils.ts]:
```suggestion
   * Gives the path to project-config.json inside a project folder
```

File: src/classes/proj-config-utils.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+   * 
+   * @param outerFolder outer folder of project (check path-utils for more info)
+   */
+  private static outerToFile(outerFolder : string) : string {
```
Comment (jarinykanen) [inline src/classes/proj-config-utils.ts]:
```suggestion
  private static outerToFile(outerFolder: string): string {
```

File: src/commands/new-repo.ts [hunk 10]
```diff
@@ -100,15 +101,18 @@ async function action(args) {
     const copy : string = await OsUtils.getCommand("copy");
```
Comment (jarinykanen) [inline src/commands/new-repo.ts]:
```suggestion
    const copy: string = await OsUtils.getCommand("copy");
```

File: src/interfaces/types.ts [hunk 11]
```diff
@@ -66,6 +66,10 @@ export interface UserConfigJson {
   osPref : string
 }
 
+export interface ProjConfigJson {
```
Comment (jarinykanen) [inline src/interfaces/types.ts:74]: Add docs

**Review by jarinykanen — CHANGES_REQUESTED** [review]

**Review by veikkahonkanen — APPROVED** [review]

File: src/interfaces/types.ts [hunk 12]
```diff
@@ -66,6 +66,15 @@ export interface UserConfigJson {
   osPref : string
 }
 
+/**
+ * Format for project-config.jsons
+ * 
+ * @property projectName name of the project
+ */
+export interface ProjConfigJson {
+  projectName : string
```
Comment (jarinykanen) [inline src/interfaces/types.ts]:
```suggestion
  projectName: string
```

**Review by jarinykanen — DISMISSED** [review]

File: src/classes/proj-config-utils.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * Exports functions related to manipulating project-config-jsons (mainly reading and writing)
+ */
+export class ProjConfigUtils {
+
+  /**
+   * Reads project-config.json from given path
+   * 
+   * @param outerFolder outer folder of the project (check path-utils for more info)
+   *
```
Comment (jarinykanen) [inline src/classes/proj-config-utils.ts]:
```suggestion
```

**Review by jarinykanen — CHANGES_REQUESTED** [review]

File: src/commands/new-repo.ts [hunk 14]
```diff
@@ -97,18 +98,21 @@ async function action(args) {
...
+    const copy: string = await OsUtils.getCommand("copy");
     try {
       execSync("git init", {cwd : repoPath});
-      execSync(`${copy} project-config.json ${folderPath}`, {cwd : `.${path.sep}resources`});
+      execSync(`${copy} project-config-template.json ${folderPath}${path.sep}project-config.json`, {cwd : `.${path.sep}resources`});
       execSync("git checkout -q -b develop", {cwd : repoPath});
       execSync(`${copy} README.md ${repoPath}`, {cwd : `.${path.sep}resources`});
       execSync(`git add README.md`, {cwd : repoPath});
       execSync(`git commit -q -m "first commit"`, {cwd : repoPath});
       execSync(`git push -q origin develop`, {cwd : repoPath});
```
Comment (veikkahonkanen) [inline src/commands/new-repo.ts]:
```
    spawnSync("git branch -m master develop", {cwd : repoPath});
    finishRepo(repoPath);
  } else {
    const copy : string = await OsUtils.getCommand("copy");
    const copy: string = await OsUtils.getCommand("copy");
    try {
      spawnSync("git init", {cwd : repoPath});
      spawnSync(`${copy} project-config.json ${folderPath}`, {cwd : `.${path.sep}resources`});
      spawnSync(`${copy} project-config-template.json ${folderPath}${path.sep}project-config.json`, {cwd : `.${path.sep}resources`});
      spawnSync("git checkout -q -b develop", {cwd : repoPath});
      spawnSync(`${copy} README.md ${repoPath}`, {cwd : `.${path.sep}resources`});
      spawnSync(`git add README.md`, {cwd : repoPath});
      spawnSync(`git commit -q -m "first commit"`, {cwd : repoPath});
      spawnSync(`git push -q origin develop`, {cwd : repoPath});
```

File: src/classes/proj-config-utils.ts [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  /**
+   * Gives the path to project-config.json inside a project folder
+   * 
+   * @param outerFolder outer folder of project (check path-utils for more info)
+   */
+  private static outerToFile(outerFolder: string): string {
+    return path.join(outerFolder, configFileName);
+  }
```
Comment (veikkahonkanen) [inline src/classes/proj-config-utils.ts:82]: Probably a job for path-utils instead

**Review by veikkahonkanen — APPROVED** [review]

**Review by jarinykanen — APPROVED** [review]

**Pull request merged**